### PR TITLE
Release cosmos db microsoft.document db 2022 11 15

### PR DESCRIPTION
**Pull request by kshittiz7**
## ARM API Information (Control Plane)

<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow.
</i>
Azure 1st Party Service can try out the [Shift Left](https://aka.ms/ShiftLeft) experience to initiate API design review from ADO code repo.  If you are interested, may request engineering support by filling in with the form https://aka.ms/ShiftLeftSupportForm.

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [ ] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. By default, Azure SDKs of all languages (.NET/Python/Java/JavaScript for both management-plane SDK and data-plane SDK, Go for management-plane SDK only ) MUST be refreshed with/after swagger of new version is published. If you prefer NOT to refresh any specific SDK language upon swagger updates in the current PR, please leave details with justification here.

### Contribution checklist (MS Employees Only):
- [ ] I commit to follow the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy) of "no breaking changes"
- [ ] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [ ] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist

> **Applicability**: :warning:
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs
> - Adding new properties
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following apply to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you copy the existing version into the new directory structure for first commit and then push new changes, including version updates, in separate commits. You can use OpenAPIHub to initialize the PR for adding a new version. For more details refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them.

### Breaking Change Review Checklist
If you have any breaking changes as defined in the [Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy/), request approval from the Breaking Change Review Board.

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Additional details on the process and office hours are on the [Breaking Change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

NOTE: To update API(s) in public preview for over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

### CODE REVIEW

**Comment by rkmanda** [comment]
Please go thru all the errors and warnings flagged here https://github.com/Azure/azure-rest-api-specs/pull/21333/checks?check_run_id=9652055777 and address them. Once you have addressed please remove the "ArmChangesRequested" label.

**Comment by rkmanda** [comment]
>       "get": {

security sensitive data MUST be returned with a POST action and not a get

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:2132 in a5f919a. [](commit_id = a5f919aa7db65d14199591137949f675673a3b9b, deletion_comment = False)

**Comment by rkmanda** [comment]
>       "get": {

If the response payload has security sensitive information make this a POST action

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:2171 in a5f919a. [](commit_id = a5f919aa7db65d14199591137949f675673a3b9b, deletion_comment = False)

**Comment by rkmanda** [comment]
>         "x-ms-long-running-operation": true,

specify x-ms-long-running-operation-options

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:2216 in a5f919a. [](commit_id = a5f919aa7db65d14199591137949f675673a3b9b, deletion_comment = False)

**Comment by rkmanda** [comment]
>     "ClientEncryptionKeyCreateUpdateParameters": {

all-of common type for ProxyResource?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:5591 in a5f919a. [](commit_id = a5f919aa7db65d14199591137949f675673a3b9b, deletion_comment = False)

**Comment by rkmanda** [comment]
>         "id": {

this should be a top level property in the envelope

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:7459 in a5f919a. [](commit_id = a5f919aa7db65d14199591137949f675673a3b9b, deletion_comment = False)

**Comment by rkmanda** [comment]
>         "algorithm": {

enum?

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:7494 in a5f919a. [](commit_id = a5f919aa7db65d14199591137949f675673a3b9b, deletion_comment = False)

**Comment by rkmanda** [comment]
>     "ClientEncryptionKeyGetProperties": {

PUT and GET response payloads need to have the same structure. Both of them MUST not have security sensitive data like the actual encryption keys

---
Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:5634 in a5f919a. [](commit_id = a5f919aa7db65d14199591137949f675673a3b9b, deletion_comment = False)

**Comment by ekuhlmann23** [comment]
Hi @rkmanda, we are also seeing a few errors related to previous api versions that are not easily addressable with this one. Could we get an exception for those?
For example:

Expected type object but found type string
Url: [Microsoft.DocumentDB/stable/2022-11-15/services.json#L293:31](https://github.com/Azure/azure-rest-api-specs/blob/a5f919aa7db65d14199591137949f675673a3b9b/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/services.json#L293:31)
ExampleUrl: [stable/2022-11-15/examples/CosmosDBServicesList.json#L22:46](https://github.com/Azure/azure-rest-api-specs/blob/a5f919aa7db65d14199591137949f675673a3b9b/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/examples/CosmosDBServicesList.json#L22:46)

**Comment by ekuhlmann23** [comment]
> > ```
> > "ClientEncryptionKeyGetProperties": {
> > ```
> 
> PUT and GET response payloads need to have the same structure. Both of them MUST not have security sensitive data like the actual encryption keys
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:5634 in [a5f919a](https://github.com/Azure/azure-rest-api-specs/commit/a5f919aa7db65d14199591137949f675673a3b9b). [](commit_id = [a5f919a](https://github.com/Azure/azure-rest-api-specs/commit/a5f919aa7db65d14199591137949f675673a3b9b), deletion_comment = False)

@kr-santosh

**Comment by kr-santosh** [comment]
> > ```
> > "ClientEncryptionKeyGetProperties": {
> > ```
> 
> PUT and GET response payloads need to have the same structure. Both of them MUST not have security sensitive data like the actual encryption keys
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:5634 in [a5f919a](https://github.com/Azure/azure-rest-api-specs/commit/a5f919aa7db65d14199591137949f675673a3b9b). [](commit_id = [a5f919a](https://github.com/Azure/azure-rest-api-specs/commit/a5f919aa7db65d14199591137949f675673a3b9b), deletion_comment = False)

Hi @rkmanda  all of these payloads have encryption keys in Wrapped Format. They are generated at the client and encrypted and then stored in CosmosDB backend as Database resource. We have already released the Public preview and now porting the Spec to Stable version

**Comment by kr-santosh** [comment]
> > ```
> >     "algorithm": {
> > ```
> 
> enum?
> 
> Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:7494 in [a5f919a](https://github.com/Azure/azure-rest-api-specs/commit/a5f919aa7db65d14199591137949f675673a3b9b). [](commit_id = [a5f919a](https://github.com/Azure/azure-rest-api-specs/commit/a5f919aa7db65d14199591137949f675673a3b9b), deletion_comment = False)

These are managed as Strings in the Core .NET SDK hence retaining the same here  also.

**Comment by kr-santosh** [comment]
> ClientEncryptionKeyCreateUpdateParameters

Following the structure as any other DatabaseResource we have in CosmosDB.

**Comment by kr-santosh** [comment]
> ClientEncryptionKeyCreateUpdateParameters

I did not understand. I dont see it under proxy

**Comment by kr-santosh** [comment]
> > > ```
> > > "ClientEncryptionKeyGetProperties": {
> > > ```
> > 
> > 
> > PUT and GET response payloads need to have the same structure. Both of them MUST not have security sensitive data like the actual encryption keys
> > Refers to: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json:5634 in [a5f919a](https://github.com/Azure/azure-rest-api-specs/commit/a5f919aa7db65d14199591137949f675673a3b9b). [](commit_id = [a5f919a](https://github.com/Azure/azure-rest-api-specs/commit/a5f919aa7db65d14199591137949f675673a3b9b), deletion_comment = False)
> 
> @kr-santosh

I have responded.

**Comment by kr-santosh** [comment]
> ClientEncryptionKeyGetProperties

They refer to ClientEncryptionKeyResource. As mentioned earlier they all have encryption keys in wrapped/encrypted form. They are encrypted/decrypted on the client.

**Comment by kr-santosh** [comment]
> get

The request is sent via client to retrieve the encrypted keys from the backend to be used post decryption on the client side. No sensitive data is exposed.

**Comment by pjohari-ms** [comment]
@weidongxu-microsoft Regarding the Avocado check, we haven't made any changes to the 2016 API version that the check is failing on. Could you take a look at it?

**Comment by pjohari-ms** [comment]
@weidongxu-microsoft Also the lintdiff errors are not for the changes that were made in this PR. They point to code that was carried forward from previous API version. These errors were not triggered before when these API versions were merged. Could you take a look at them as well?

**Comment by weidongxu-microsoft** [comment]
@pjohari-ms Regardless of whether it is caused by your change or not, all of them is required to be fix or at least suppressed.

**Comment by ekuhlmann23** [comment]
> @pjohari-ms Regardless of whether it is caused by your change or not, all of them is required to be fix or at least suppressed.
> 
> Some of them are new rules that wasn't there on previous commit to the spec.

We are still having issues with the Avocado check, though. Not sure why this error ([MISSING_APIS_IN_DEFAULT_TAG](https://github.com/Azure/avocado/blob/master/README.md#MISSING_APIS_IN_DEFAULT_TAG)) is popping up here. It seems like an outdated rule.

**Comment by weidongxu-microsoft** [comment]
> We are still having issues with the Avocado check, though. Not sure why this error ([MISSING_APIS_IN_DEFAULT_TAG](https://github.com/Azure/avocado/blob/master/README.md#MISSING_APIS_IN_DEFAULT_TAG)) is popping up here. It seems like an outdated rule.

It seems to say some API in [Microsoft.DocumentDB/stable/2016-03-31/cosmos-db.json](https://github.com/Azure/azure-rest-api-specs/blob/2ee9bddecbf619e7250d938cf61bb5559832c6d4/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2016-03-31/cosmos-db.json) does not exist in your latest [cosmos-db.json](https://github.com/Azure/azure-rest-api-specs/blob/2ee9bddecbf619e7250d938cf61bb5559832c6d4/specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/cosmos-db.json)

Is these API get deleted?

If not, maybe it is a bug in tooling.

**Comment by pjohari-ms** [comment]
@weidongxu-microsoft The API isn't deleted, it's just a very old API. We have also got an exception during our last API version release for Avocado, and were told that this won't happen for future PRs. Seems like we need to get an exception again for this PR?

**Comment by weidongxu-microsoft** [comment]
OK, you can follow this to get it suppressed https://github.com/Azure/azure-rest-api-specs/blob/main/documentation/ci-fix.md#suppression-process

Note you've got a new LintDiff error

**Review by pjohari-ms — APPROVED** [review]

**Comment by pjohari-ms** [comment]
@ruowan We requested avocado approval in this PR: https://github.com/Azure/azure-rest-api-specs/pull/20419 . This current PR is also failing the avocado check (for an older API version) and I have requested a suppression for it. Could you please provide an approval for avocado in the meantime?

**Comment by pjohari-ms** [comment]
@weidongxu-microsoft @ruowan anything on the Avocado exception? I have already requested a suppression but there doesn't seem to be any progress on that. Can you approve the avocado exception for this PR for now?

**Comment by weidongxu-microsoft** [comment]
You've not get ARM approval yet. Why worry about Avocado now.

**Comment by pjohari-ms** [comment]
@rkmanda We have added the response to your comments. Could you take a look?

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/services.json
Comment (sjanamma) [inline specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/services.json:46]: Can v4 common types be used in case it's not a breaking change?

Comment (sjanamma) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/services.json:46]: V4 mostly has changes to support RPaaS RPs. We recommend you going with the latest version possible.

**Comment by sjanamma** [comment]
We still see some open comments. Please address on how you are planning to handle them.

File: specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/services.json
Comment (pjohari-ms) [reply specification/cosmos-db/resource-manager/Microsoft.DocumentDB/stable/2022-11-15/services.json:46]: Noted. We'll take this up in the next API version changes.

**Comment by pjohari-ms** [comment]
> We still see some open comments. Please address on how you are planning to handle them.
Please let us know if there are any more comments remaining for this PR

**Comment by pjohari-ms** [comment]
@weidongxu-microsoft @ruowan We have an ARM signoff now. Can you approve the avocado exception for this PR for now?

**Comment by pjohari-ms** [comment]
@weidongxu-microsoft since you're the PR assignee as well, do we need anything else before we can merge this PR? If not, can you please merge this PR as well?

**Pull request merged**
